core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala

```diff
@@ -621,6 +615,80 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
   }
 }
 
-private[spark] object CoarseGrainedSchedulerBackend {
+private[spark] object CoarseGrainedSchedulerBackend extends Logging {
   val ENDPOINT_NAME = "CoarseGrainedScheduler"
+
+  // abort TaskSetManager without exception
+  private def abortTaskSetManager(
+      scheduler: TaskSchedulerImpl,
+      taskId: Long,
+      msg: => String,
+      exception: Option[Throwable] = None): Unit = scheduler.synchronized {
+    scheduler.taskIdToTaskSetManager.get(taskId).foreach { taskSetMgr =>
+      try {
+        taskSetMgr.abort(msg, exception)
+      } catch {
+        case e: Exception => logError("Exception in error callback", e)
+      }
+    }
+  }
+
+  private def isZombieTaskSetManager(
+    scheduler: TaskSchedulerImpl,
+    taskId: Long): Unit = scheduler.synchronized {
+    !scheduler.taskIdToTaskSetManager.get(taskId).exists(_.isZombie)
+  }
+
+  private def getTaskSetManager(
+    scheduler: TaskSchedulerImpl,
+    taskId: Long): Option[TaskSetManager] = scheduler.synchronized {
+    scheduler.taskIdToTaskSetManager.get(taskId)
+  }
+
+  private[scheduler] def prepareSerializedTask(
```

Comment:
@kayousterhout 
@squito 
Refactoring code in https://github.com/apache/spark/pull/17116